src/zip_code_tree.hpp

```diff
@@ -405,114 +426,231 @@ class ZipCodeForest {
+    //TODO: I think the distance_limit should just be the same as the gap_distance_limit
+    /// If a distance_limit is given, then distances larger than the distance limit are not
+    /// guaranteed to be accurate, but will be greater than the distance_limit
+    template<typename Minimizer>
+    void fill_in_forest(const vector<Seed>& seeds, const VectorView<Minimizer>& minimizers, 
+                      const SnarlDistanceIndex& distance_index,
+                      size_t gap_distance_limit,
                       size_t distance_limit = std::numeric_limits<size_t>::max());
+
     private:
-    //The seeds that are taken as input
-    //The order of the seeds will never change, but the vector is not const because the zipcodes
-    //decoders may change
-    const vector<Seed>* seeds;
 
-    public:
 
-    void print_self() const {
-        for (size_t i = 0 ; i < trees.size() ; i++) {
-            const auto& tree = trees[i];
-            cerr << i << ": ";
-            tree.print_self();
-        }
-    }
-    void validate_zip_forest(const SnarlDistanceIndex& distance_index, size_t distance_limit=std::numeric_limits<size_t>::max()) const;
+    /***********************************************************************************************
+
+      Data structures and helper functions for construction
+
+      **********************************************************************************************
+
+      Construction is done in a depth-first traversal of the snarl tree. So when each 
+      snarl tree node is visited, the start of the structure is added to the zip tree, then each of 
+      its children is added to the zip tree, then the end of the structure is added.
+      
+      The traversal of the snarl tree is accomplished by progressively sorting the seeds to identify
+      the snarl tree structures that they lie on. Using the zip codes, the seeds can be sorted at 
+      each depth separately. The seeds get sorted using a radix-like sort, starting with the root of 
+      the snarl tree and moving down. So first, the seeds are sorted into connected components. The 
+      components are saved as "intervals" that remember the range in the sort order that the seeds 
+      occur on. Each interval of seeds represents a root-level snarl or chain. Each interval is then
+      sorted to order the seeds along the snarl or chain, and new intervals are found representing 
+      ranges of seeds on the children.
+
+      Each snarl and chain is comprised of the start and end bounds, the children, and distances 
+      between children/bounds. So as each child is added, we will need to know what came before it
+      in the parent snarl/chain so that we can add the distances. We also need to remember the 
+      ancestors of each snarl and chain as we are building them, so that we can close each structure
+      properly. All of this information is stored in a forest_growing_state_t as the zip trees are 
```

Comment:
I wonder if instead of passing around a state to everything, it might be better to make some kind of builder class and have it own the state, to reduce the number of arguments to all the functions? Or does it need to be like this for testability?

A builder class might also help you break the build code out into its own file which might be clearer?